client/data/charges/types.ts

```diff
@@ -0,0 +1,282 @@
+/* eslint-disable camelcase */
+export interface Address {
```

Comment:
I thought about doing that, but I don't think that's a good idea because we tend to restructure the data somewhat on the server side. While we don't do this with `charges` I think it's a good enough reason to define this ourselves.

I would additionally say that the typing I've laid out for the payment method details is slightly more robust, but that's a minor nitpick 🙂

Reply:
Conceptually they aren't Stripe types, they're WCPay types. So to me it makes sense to create our own, as we have in the server-side code.

However, pragmatically I see @marcinbot's point. I don't have strong feelings either way. I'd perhaps suggest we go with whatever is going to be easiest to keep updated and maintained?

Reply:
> I'd perhaps suggest we go with whatever is going to be easiest to keep updated and maintained?

I honestly think both methods are about the same.

Using the Stripe object will mean we need to keep the Stripe JS up to date, or at least in line with the version of the Stripe API we're using.

Maintaining our own will require small updates to the JS when Stripe updates their API, just like it would if we were relying on the object from Stripe.

I think this really just comes down to our preference 🤷

Reply:
I'm leaning towards @marcinbot's suggestion. It's easier to start with Stripe's types and adjust where needed (if needed). 

> Conceptually they aren't Stripe types, they're WCPay types. So to me it makes sense to create our own, as we have in the server-side code.

I like that point too, so to make it a bit more robust we could create own interfaces/types which do empty extend of Stripe interfaces ( e.g. `WCPayInterface extends StripeInterface {}` ) for start and adjust only where needed. This approach would allow us to stick to own namespace but get many type definitions from Stripe for low cost.

Reply:
> we could create own interfaces/types which do empty extend of Stripe interfaces ( e.g. WCPayInterface extends StripeInterface {} ) for start and adjust only where needed.

I love this approach. Let's do it!

> the maintenance costs might be similar, but the cost of adding a feature is greater.

You make an excellent point there. I agree; adding new features/types will be exponentially more difficult and involved.

In any case, I agree @vbelolapotkov's suggestion here is the best way forward. Thank you all for chiming in here! 🙂

Reply:
One problem here is that Stripe doesn't export all the types they declare, e.g. the `Card` details [used in the `PaymentMethod`](https://github.com/stripe/stripe-js/blob/db51d9576a66ff0de4326aeac19dc73934935bc8/types/api/PaymentMethods.d.ts#L100-L150) are not exported, but a [`Card` for card details is](https://github.com/stripe/stripe-js/blob/db51d9576a66ff0de4326aeac19dc73934935bc8/types/api/Cards.d.ts#L5-L125). I'm not sure what the difference is.

They also declare a `BillingDetails` object, [but don't export it](https://github.com/stripe/stripe-js/blob/db51d9576a66ff0de4326aeac19dc73934935bc8/types/api/PaymentMethods.d.ts#L78-L98), so we can't import it explicitly 🤔 

Another issue; [they don't declare a `Charge` type at all](https://github.com/stripe/stripe-js/tree/v1.12.1/types/api). I'll do my best with what we have, but we're clearly going to have to manage many of our own types.

Reply:
Thanks for digging into this @reykjalin ! It's a bit disappointing the types are not exported, however, it's still feels like a good idea to copy or use Stripe's src code as a reference.

Reply:
I've made a pretty big commit that adds some types from `stripe-node` I think we use to `data/stripe`. It's not complete; if there was a type I don't think we use I replaced that with `unknown` to keep the commit as small as possible.

I used whatever I found in `stripe-js` to fill in the types, but there were a significant amount of interfaces missing. All of those interfaces are in `data/stripe`.

All of this to say, there's no need to review things in `client/data/stripe`; those are just copied from `stripe-node` and cleaned up a bit to fit WCPay, e.g. use available types from `stripe-js`, removing empty lines and comments, and removing namespaces.

Reply:
Yeah, that's what I've been going with; only prefixing a type if we modify it. And really, we should extend it outside `data/stripe`, since I think maintaining that separation between stripe things and our things will be good.

Take `WCPayCharge` as an example; I added the `Charge` interface (copied from Stripe) to `data/stripe/charges.d.ts`, and then extended the interface to be a bit nicer to work with in `data/charges/types.ts`. I think that clearly maintains the separation of "this is the Stripe type, and that is our type".